### PR TITLE
map_variables

### PR DESCRIPTION
**Pull request by cooijmanstim**
(Work in progress on #3177 , do not merge)

Now the replacement descends into scan nodes. For some reason that I don't understand, I need to keep track of those scan nodes that I have already replaced. Otherwise the local optimizer will be called again and again on the replaced node. Changing the order from out_to_in to in_to_out doesn't change that.

I'll sort out the mess of imports later.

(The other pull request was from my master and got contaminated with unrelated commits; I'll use this one from now on.)

### CODE REVIEW

**Comment by carriepl** [comment]
If the other pull request is not used anymore, you should close it.

**Comment by carriepl** [comment]
Also, it's good that it now works for scan nodes but it should also work for OpFromGraph nodes (which is just a way of bundling a computational graph into a single op

File: theano/gof/graph.py [hunk 1]
```diff
@@ -816,6 +816,92 @@ def clone_get_equiv(inputs, outputs, copy_inputs_and_orphans=True, memo=None):
...
+    fg = FunctionGraph(inputs_, graphs, clone=False)
+
+    nodes_seen = set()
+
+    @local_optimizer(None)
+    def local_transform(node):
+        if node in nodes_seen:
+            return False
+        # FIXME: replacing inputs won't work because they are not
+        # outputs of any Apply node
```
Comment (carriepl) [inline theano/gof/graph.py]: If you want to support input replacements, you can do it after the optimization with your 'local_transform', loop over the inputs of the optimized graph and get their replacements if applicable and then perform a clone with replacement.

**Comment by carriepl** [comment]
Some tests should also be implemented to test this feature.

**Comment by carriepl** [comment]
Can you add a test for an input replacement inside either a Scan or OpFromGraph? Once that's done, I'm fine with merging as long as Travis is happy.

**Comment by carriepl** [comment]
And travis seems unhappy again  : 
```
ERROR: theano.gof.tests.test_graph.TestMapVariables.test_opfromgraph

----------------------------------------------------------------------

Traceback (most recent call last):

  File "/home/travis/miniconda/envs/pyenv/lib/python3.3/site-packages/nose-1.3.4-py3.3.egg/nose/case.py", line 198, in runTest

    self.test(*self.arg)

  File "/home/travis/miniconda/envs/pyenv/lib/python3.3/site-packages/theano/gof/tests/test_graph.py", line 199, in test_opfromgraph

    for m, n in itertools.combinations(xrange(10), 2):

NameError: global name 'xrange' is not defined
```

**Comment by carriepl** [comment]
Ok, I'll wait for that modification too. You know where to put it?

**Comment by carriepl** [comment]
I guess putting it in the same place as 'clone' would probably make the most sense, given that both have similar uses.

**Comment by carriepl** [comment]
Travis is complaining because of a syntax error

**Comment by carriepl** [comment]
Why was this moved to scan_utils? Was there something preventing it from being in the same place as clone? Since this is useful for more than Scan, I'm not sure that files only related to Scan is the best place for it.

File: theano/scan_module/scan_utils.py [hunk 2]
```diff
@@ -256,6 +256,208 @@ def clone(output,
...
+        if not constants and not foreign_inputs:
+            return new_graph
+
+        replacements = []
+
+        # constants just need to be replaced by copies that the inner
+        # `fg` can take ownership of
+        for input_ in constants:
+            new_input = input_.clone()
+            new_input.name = "%s_copiedd" % new_input.name
```
Comment (carriepl) [inline theano/scan_module/scan_utils.py]: 'copiedd'

**Comment by cooijmanstim** [comment]
Yes, I moved it to scan_utils because there is a "clone" there.  This clone is the one that does replacement.  Both would be better moved elsewhere if you ask me.

I'll fix the syntax error and the typos when I get back.

**Comment by carriepl** [comment]
I talked with @nouiz and, for this PR, map_variables can be left in scan_utils. It will probably be moved eventually but it's not obvious to him, at the moment, where it should go.

File: theano/scan_module/tests/test_scan_utils.py [hunk 3]
```diff
@@ -11,3 +15,146 @@ def test_equal_compuations():
...
+            s2, = map_variables(self.replacer, [s])
+        except NotImplementedError as e:
+            e = sys.exc_info()[1]
+            assert("introduces shared variable" in str(e))
+            return
+
+        # test failed
+        return 0
+
+    def test_scan_with_shared_update3(self):
```
Comment (carriepl) [inline theano/scan_module/tests/test_scan_utils.py]: 3? Is there supposed to be a test_scan_with_shared_update2 before this test?

File: theano/scan_module/tests/test_scan_utils.py [hunk 4]
```diff
@@ -11,3 +15,146 @@ def test_equal_compuations():
...
+                           outputs_info=[numpy.array(0.)])
+        try:
+            s2, = map_variables(self.replacer, [s])
+        except NotImplementedError as e:
+            e = sys.exc_info()[1]
+            assert("introduces shared variable" in str(e))
+            return
+
+        # test failed
+        return 0
```
Comment (carriepl) [inline theano/scan_module/tests/test_scan_utils.py]: You can replace, from the try to the return, by a call to self.assertRaises(NotImplementedError, map_variables, self.replacer, [s]).
For self.assertRaises to be defined, your test class must inherit from unittest.TestCase

This comment also applies to the other test that follows the same pattern.

**Comment by carriepl** [comment]
Originally, I asked for a test adding inputs to either a Scan or OpFromGraph but given the complexity that this PR has reached, I would prefer if we had both. You've already made a test for Scan, can you add a small one for OpFromGraph? After this and my other comment, I think this will be all for this PR.

File: theano/scan_module/tests/test_scan_utils.py (hunk 3 above)
Comment (cooijmanstim) [inline theano/scan_module/tests/test_scan_utils.py]: Good call; yes there was but I merged it into test_scan_with_shared_update.

**Comment by cooijmanstim** [comment]
It took some time to get the new test_opfromgraph to pass, but it does now that I added additional_inputs=[outer, shared] to the calls to map_variables().  It's interesting that I didn't have to do this for the scan tests.

**Comment by carriepl** [comment]
Travis is not done running but it already spotted a flake8 issue : 
/home/travis/miniconda/envs/pyenv/lib/python2.6/site-packages/theano/scan_module/tests/test_scan_utils.py:1:1: F401 'sys' imported but unused

**Comment by carriepl** [comment]
Good work. This will probably be expended and improved as it becomes more heavily used. For now, this seems solid and there's a good list of tests. I think this is good for merge.

**Pull request merged**